File: internal/search/temporal.go
package search

import (
	"math"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/eleboucher/memini/internal/store"
)

// Temporal targeting boosts candidates dated near the time a query references
// ("what did I do three weeks ago") rather than near now, which the monotonic
// recency factor cannot do. It only fires when the query names a relative time.
//
// DefaultTemporalBoost is the maximum relevance amplification a perfectly
// on-target candidate gets (score × (1 + boost) at zero distance); it ramps to
// 0 by 3x the tolerance. Multiplicative so proximity can reorder comparably
// relevant candidates without letting a date-near but weakly relevant memory
// displace a strong match.
const DefaultTemporalBoost = 0.40

// TimeAnchor is a resolved relative-time reference: an answer is expected
// roughly Days before the query's "now", give or take Tolerance days.
type TimeAnchor struct {
	Days      int
	Tolerance int
}

// AnchorExtractor resolves a query's time reference, if any, against the
// query's "now" (absolute references like "in March" need it). The regex
// implementation is the no-LLM default; an LLM extractor (looser phrasing like
// "a couple weeks before my trip") can plug in via the same interface.
type AnchorExtractor interface {
	Anchor(query string, now time.Time) (TimeAnchor, bool)
}

// RegexAnchorExtractor matches common English time phrases — relative ("3
// weeks ago") and absolute ("in March", "March 14th", "last summer", "in
// 2025") — firing only on templated expressions.
type RegexAnchorExtractor struct{}

var temporalPatterns = []struct {
	re        *regexp.Regexp
	days, tol int
	numMul    int // when != 0, multiply the captured integer by this
}{
	{regexp.MustCompile(`(\d+)\s+days?\s+ago`), 0, 2, 1},
	{regexp.MustCompile(`a\s+couple\s+(?:of\s+)?days?\s+ago`), 2, 2, 0},
	{regexp.MustCompile(`yesterday`), 1, 1, 0},
	{regexp.MustCompile(`(\d+)\s+weeks?\s+ago`), 0, 5, 7},
	{regexp.MustCompile(`(?:a|last)\s+week\b`), 7, 3, 0},
	{regexp.MustCompile(`(\d+)\s+months?\s+ago`), 0, 10, 30},
	{regexp.MustCompile(`(?:a|last)\s+month\b`), 30, 7, 0},
	{regexp.MustCompile(`(?:a|last)\s+year\b`), 365, 30, 0},
	{regexp.MustCompile(`recently`), 14, 14, 0},
}

// Anchor implements AnchorExtractor. Relative phrases resolve without now;
// absolute references (month, month+day, season, year) resolve to the most
// recent past occurrence relative to now and are skipped when now is zero.
func (RegexAnchorExtractor) Anchor(query string, now time.Time) (TimeAnchor, bool) {
	q := strings.ToLower(query)
	for _, p := range temporalPatterns {
		m := p.re.FindStringSubmatch(q)
		if m == nil {
			continue
		}
		if p.numMul != 0 && len(m) > 1 {
			n, err := strconv.Atoi(m[1])
			if err != nil {
				continue
			}
			return TimeAnchor{Days: n * p.numMul, Tolerance: p.tol}, true
		}
		return TimeAnchor{Days: p.days, Tolerance: p.tol}, true
	}
	if now.IsZero() {
		return TimeAnchor{}, false
	}
	return absoluteAnchor(q, now)
}

var monthsByName = map[string]time.Month{
	"january": time.January, "february": time.February, "march": time.March,
	"april": time.April, "may": time.May, "june": time.June, "july": time.July,
	"august": time.August, "september": time.September, "october": time.October,
	"november": time.November, "december": time.December,
}

const monthAlternation = `(january|february|march|april|may|june|july|august|september|october|november|december)`

var (
	// "October 13, 2023" — a fully-specified date.
	monthDayYearPattern = regexp.MustCompile(`\b` + monthAlternation + `\s+(\d{1,2})(?:st|nd|rd|th)?,?\s+((?:19|20)\d{2})\b`)
	// "July 2023" — a month in an explicit year.
	monthYearPattern = regexp.MustCompile(`\b` + monthAlternation + `\s+((?:19|20)\d{2})\b`)
	// "March 14", "on march 14th" — a specific day.
	monthDayPattern = regexp.MustCompile(`\b` + monthAlternation + `\s+(\d{1,2})(?:st|nd|rd|th)?\b`)
	// "in March", "last march", "back in march", "since march" — a month. The
	// leading preposition keeps bare month mentions ("May I ask") from firing.
	monthPattern = regexp.MustCompile(`\b(?:in|last|back in|since|during)\s+` + monthAlternation + `\b`)
	// "last summer", "in the winter", "this past fall" — a season (northern
	// mid-season dates; the tolerance absorbs the hemisphere ambiguity).
	seasonPattern = regexp.MustCompile(`\b(?:last|in the|this past)\s+(spring|summer|fall|autumn|winter)\b`)
	// "in 2025" — a year.
	yearPattern = regexp.MustCompile(`\bin\s+((?:19|20)\d{2})\b`)
)

// seasonMidMonth anchors each season at its northern mid-point.
var seasonMidMonth = map[string]time.Month{
	"spring": time.April, "summer": time.July,
	"fall": time.October, "autumn": time.October, "winter": time.January,
}

// absoluteAnchor resolves date references, most specific first: an explicit
// year pins the target exactly; without one, the most recent past occurrence
// relative to now wins.
func absoluteAnchor(q string, now time.Time) (TimeAnchor, bool) {
	if m := monthDayYearPattern.FindStringSubmatch(q); m != nil {
		day, errD := strconv.Atoi(m[2])
		year, errY := strconv.Atoi(m[3])
		if errD == nil && errY == nil && day >= 1 && day <= 31 {
			target := time.Date(year, monthsByName[m[1]], day, 0, 0, 0, 0, time.UTC)
			return TimeAnchor{Days: daysAgo(target, now), Tolerance: 2}, true
		}
	}
	if m := monthYearPattern.FindStringSubmatch(q); m != nil {
		if year, err := strconv.Atoi(m[2]); err == nil {
			target := time.Date(year, monthsByName[m[1]], 15, 0, 0, 0, 0, time.UTC)
			return TimeAnchor{Days: daysAgo(target, now), Tolerance: 15}, true
		}
	}
	if m := monthDayPattern.FindStringSubmatch(q); m != nil {
		day, err := strconv.Atoi(m[2])
		if err == nil && day >= 1 && day <= 31 {
			target := recentOccurrence(monthsByName[m[1]], day, now)
			return TimeAnchor{Days: daysAgo(target, now), Tolerance: 2}, true
		}
	}
	if m := monthPattern.FindStringSubmatch(q); m != nil {
		target := recentOccurrence(monthsByName[m[1]], 15, now)
		return TimeAnchor{Days: daysAgo(target, now), Tolerance: 15}, true
	}
	if m := seasonPattern.FindStringSubmatch(q); m != nil {
		target := recentOccurrence(seasonMidMonth[m[1]], 15, now)
		return TimeAnchor{Days: daysAgo(target, now), Tolerance: 45}, true
	}
	if m := yearPattern.FindStringSubmatch(q); m != nil {
		year, err := strconv.Atoi(m[1])
		if err == nil {
			target := time.Date(year, time.July, 1, 0, 0, 0, 0, time.UTC)
			return TimeAnchor{Days: daysAgo(target, now), Tolerance: 120}, true
		}
	}
	return TimeAnchor{}, false
}

// recentOccurrence is the most recent month/day occurrence at or before now
// (this year when already past, otherwise last year).
func recentOccurrence(month time.Month, day int, now time.Time) time.Time {
	t := time.Date(now.Year(), month, day, 0, 0, 0, 0, time.UTC)
	if t.After(now) {
		t = t.AddDate(-1, 0, 0)
	}
	return t
}

func daysAgo(target, now time.Time) int {
	return int(now.Sub(target).Hours() / 24)
}

// RerankTemporal re-ranks with the composite weights, then — when ex resolves a
// relative-time reference in query — scales each score by a date-proximity
// factor toward (now - anchor) so a candidate dated near the referenced time can climb past a
// marginally-more-similar one. With no time reference (or no extractor) it is
// exactly RerankWith. boost <= 0 also degrades to plain composite re-rank.
func RerankTemporal(
	results []store.Scored, query string, now time.Time, w RerankWeights, ex AnchorExtractor, boost float64,
) []store.Scored {
	base := RerankWith(results, now, w)
	if ex == nil || boost <= 0 || now.IsZero() {
		return base
	}
	anchor, ok := ex.Anchor(query, now)
	if !ok {
		return base
	}
	target := now.AddDate(0, 0, -anchor.Days)

	type ranked struct {
		sc  store.Scored
		val float64
		pos int
	}
	out := make([]ranked, len(base))
	for i, r := range base {
		// A backdated fact's ValidFrom is when it was true — that, not the
		// row's write date, is what a time reference points at.
		date := r.Memory.CreatedAt
		if r.Memory.ValidFrom != nil {
			date = *r.Memory.ValidFrom
		}
		// Multiplicative, not additive: proximity amplifies a candidate's own
		// relevance (up to +boost×100%) rather than adding a flat bonus. A
		// date-near memory that is also relevant rises; a date-near but weakly
		// relevant one has little score to amplify, so it cannot leapfrog a
		// strong match on date alone — temporal targeting reorders among
		// comparable candidates instead of displacing relevance.
		val := r.Score * (1 + boost*temporalProximity(date, target, anchor.Tolerance))
		sc := r
		sc.Score = val
		out[i] = ranked{sc: sc, val: val, pos: i}
	}
	sort.SliceStable(out, func(i, j int) bool {
		if out[i].val != out[j].val {
			return out[i].val > out[j].val
		}
		return out[i].pos < out[j].pos
	})
	reranked := make([]store.Scored, len(out))
	for i, r := range out {
		reranked[i] = r.sc
	}
	return reranked
}

// temporalProximity is 1 when date is within tol days of target, ramps linearly
// to 0 at 3*tol, and is 0 beyond that or when the date is unknown.
func temporalProximity(date, target time.Time, tol int) float64 {
	if date.IsZero() || tol <= 0 {
		return 0
	}
	delta := math.Abs(target.Sub(date).Hours() / 24)
	switch {
	case delta <= float64(tol):
		return 1
	case delta <= float64(3*tol):
		return 1 - (delta-float64(tol))/float64(2*tol)
	default:
		return 0
	}
}

File: crates/memini-core/src/temporal.rs
use crate::search::{RerankWeights, Scored, rerank_with};
use chrono::{DateTime, Datelike, Duration, NaiveDate, TimeZone, Utc};
use regex::Regex;
use std::sync::LazyLock;

pub const DEFAULT_TEMPORAL_BOOST: f64 = 0.40;
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct TimeAnchor {
    pub days: i64,
    pub tolerance: i64,
}
static NUMBERED: LazyLock<Vec<(Regex, i64)>> = LazyLock::new(|| {
    vec![
        (Regex::new(r"(\d+)\s+days?\s+ago").unwrap(), 1),
        (Regex::new(r"(\d+)\s+weeks?\s+ago").unwrap(), 7),
        (Regex::new(r"(\d+)\s+months?\s+ago").unwrap(), 30),
    ]
});
static MONTH_DAY_YEAR: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(january|february|march|april|may|june|july|august|september|october|november|december)\s+(\d{1,2})(?:st|nd|rd|th)?,?\s+((?:19|20)\d{2})\b").unwrap()
});
static MONTH_YEAR: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(january|february|march|april|may|june|july|august|september|october|november|december)\s+((?:19|20)\d{2})\b").unwrap()
});
static MONTH_DAY: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(january|february|march|april|may|june|july|august|september|october|november|december)\s+(\d{1,2})(?:st|nd|rd|th)?\b").unwrap()
});
static MONTH_ONLY: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(?:in|last|back in|since|during)\s+(january|february|march|april|may|june|july|august|september|october|november|december)\b").unwrap()
});
static SEASON: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(r"\b(?:last|in the|this past)\s+(spring|summer|fall|autumn|winter)\b").unwrap()
});
static YEAR: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"\bin\s+((?:19|20)\d{2})\b").unwrap());

pub fn anchor(query: &str, now: DateTime<Utc>) -> Option<TimeAnchor> {
    let query = query.to_lowercase();
    for (pattern, multiplier) in NUMBERED.iter() {
        if let Some(captures) = pattern.captures(&query) {
            return Some(TimeAnchor {
                days: captures[1].parse::<i64>().ok()? * multiplier,
                tolerance: if *multiplier == 1 {
                    2
                } else if *multiplier == 7 {
                    5
                } else {
                    10
                },
            });
        }
    }
    for (phrase, days, tolerance) in [
        ("a couple days ago", 2, 2),
        ("a couple of days ago", 2, 2),
        ("yesterday", 1, 1),
        ("a week", 7, 3),
        ("last week", 7, 3),
        ("a month", 30, 7),
        ("last month", 30, 7),
        ("a year", 365, 30),
        ("last year", 365, 30),
        ("recently", 14, 14),
    ] {
        if query.contains(phrase) {
            return Some(TimeAnchor { days, tolerance });
        }
    }
    if let Some(c) = MONTH_DAY_YEAR.captures(&query) {
        return date_anchor(
            c[3].parse().ok()?,
            month(&c[1])?,
            c[2].parse().ok()?,
            now,
            2,
        );
    }
    if let Some(c) = MONTH_YEAR.captures(&query) {
        return date_anchor(c[2].parse().ok()?, month(&c[1])?, 15, now, 15);
    }
    if let Some(c) = MONTH_DAY.captures(&query) {
        return recent_anchor(month(&c[1])?, c[2].parse().ok()?, now, 2);
    }
    if let Some(c) = MONTH_ONLY.captures(&query) {
        return recent_anchor(month(&c[1])?, 15, now, 15);
    }
    if let Some(c) = SEASON.captures(&query) {
        let month = match &c[1] {
            "spring" => 4,
            "summer" => 7,
            "fall" | "autumn" => 10,
            "winter" => 1,
            _ => return None,
        };
        return recent_anchor(month, 15, now, 45);
    }
    if let Some(c) = YEAR.captures(&query) {
        return date_anchor(c[1].parse().ok()?, 7, 1, now, 120);
    }
    None
}
fn month(value: &str) -> Option<u32> {
    [
        "january",
        "february",
        "march",
        "april",
        "may",
        "june",
        "july",
        "august",
        "september",
        "october",
        "november",
        "december",
    ]
    .iter()
    .position(|v| *v == value)
    .map(|v| v as u32 + 1)
}
fn date_anchor(
    year: i32,
    month: u32,
    day: u32,
    now: DateTime<Utc>,
    tolerance: i64,
) -> Option<TimeAnchor> {
    let target =
        Utc.from_utc_datetime(&NaiveDate::from_ymd_opt(year, month, day)?.and_hms_opt(0, 0, 0)?);
    Some(TimeAnchor {
        days: (now - target).num_hours() / 24,
        tolerance,
    })
}
fn recent_anchor(month: u32, day: u32, now: DateTime<Utc>, tolerance: i64) -> Option<TimeAnchor> {
    let mut year = now.year();
    let mut target =
        Utc.from_utc_datetime(&NaiveDate::from_ymd_opt(year, month, day)?.and_hms_opt(0, 0, 0)?);
    if target > now {
        year -= 1;
        target =
            Utc.from_utc_datetime(&NaiveDate::from_ymd_opt(year, month, day)?.and_hms_opt(0, 0, 0)?)
    }
    Some(TimeAnchor {
        days: (now - target).num_hours() / 24,
        tolerance,
    })
}
pub fn rerank_temporal(
    results: &[Scored],
    query: &str,
    now: DateTime<Utc>,
    weights: RerankWeights,
    stability_k: f64,
    boost: f64,
) -> Vec<Scored> {
    let mut base = rerank_with(results, now, weights, stability_k);
    let Some(anchor) = anchor(query, now).filter(|_| boost > 0.0) else {
        return base;
    };
    let target = now - Duration::days(anchor.days);
    for item in &mut base {
        let date = item.memory.valid_from.unwrap_or(item.memory.created_at);
        item.score *= 1.0 + boost * proximity(date, target, anchor.tolerance);
    }
    base.sort_by(|a, b| b.score.total_cmp(&a.score));
    base
}
fn proximity(date: DateTime<Utc>, target: DateTime<Utc>, tolerance: i64) -> f64 {
    if tolerance <= 0 {
        return 0.0;
    }
    let delta = ((target - date).num_seconds() as f64 / 86400.0).abs();
    if delta <= tolerance as f64 {
        1.0
    } else if delta <= (3 * tolerance) as f64 {
        1.0 - (delta - tolerance as f64) / (2 * tolerance) as f64
    } else {
        0.0
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn extracts_relative_and_absolute_anchors() {
        let now = Utc.with_ymd_and_hms(2026, 6, 1, 0, 0, 0).unwrap();
        assert_eq!(
            anchor("what did I do 3 weeks ago", now),
            Some(TimeAnchor {
                days: 21,
                tolerance: 5
            })
        );
        assert_eq!(
            anchor("the milestone yesterday", now),
            Some(TimeAnchor {
                days: 1,
                tolerance: 1
            })
        );
        assert_eq!(
            anchor("what shipped back in march", now),
            date_anchor(2026, 3, 15, now, 15)
        );
        assert_eq!(
            anchor("on october 13, 2023", now),
            date_anchor(2023, 10, 13, now, 2)
        );
        assert!(anchor("may i ask", now).is_none());
    }
}
